### ISSUE TITLE
Voice Chat

### ISSUE THREAD

**Issue by ReynoldsCahoon** [azerothcore/azerothcore-wotlk#5063]
## CURRENT BEHAVIOUR:
Voice Chat ([Added in Patch 2.2.0](https://wowpedia.fandom.com/wiki/Patch_2.2.0)) support is currently missing.

## EXPECTED BLIZZLIKE BEHAVIOUR:
Players should be able to create/join voice chat channels to communicate with other players. They should also have the ability to mute other players individually.

## SOURCE:
| Channels & Moderation | Mute Players |
|---|---|
| ![image](https://user-images.githubusercontent.com/2556251/112986316-31752180-9116-11eb-9ddc-1471c22135a8.png) | ![image](https://user-images.githubusercontent.com/2556251/112986355-3e921080-9116-11eb-9fd0-d44312bd4831.png) |

| Old Voice Chat UI (Probably pre-WotLK) |
|---|
| ![image](https://user-images.githubusercontent.com/2556251/112986599-89138d00-9116-11eb-96ec-f81cf4fad356.png) |

## STEPS TO REPRODUCE THE PROBLEM:
1. Connect to AzerothCore server.
2. See that no visible voice options are available.

## EXTRA NOTES:
[TrinityCore discussion](https://github.com/TrinityCore/TrinityCore/issues/15057): References Ascent, which [had a version of this running](https://github.com/SkyFire/ascent_classic/tree/master/src/ascent-voicechat).

## AC HASH/COMMIT:
d4b8defcd8e6971bdd558b1c360de9dcca9767b3

## OPERATING SYSTEM:
Windows 10

## MODULES:
None

## CUSTOMIZATIONS:
None

**Comment by ReynoldsCahoon** [comment]
Beyond the expected Blizzlike behavior, it would be nice if this were extendable via the modular structure. An example module would be the addition of proximity voice.

**Comment by yuanf225** [comment]
Can be made into a team/team voice function, it will only take effect when joining a team/team, you can choose to turn on/mute

**Comment by ReynoldsCahoon** [comment]
> Can be made into a team/team voice function, it will only take effect when joining a team/team, you can choose to turn on/mute

This sounds like module functionality, as the original client/server supported this in a very specific way.

**Comment by yuanf225** [comment]
The text is cold, only warm words can play more emotionally ：D

**Comment by ReynoldsCahoon** [comment]
Realized that the repo contains a [VoiceChatHandler](https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Handlers/VoiceChatHandler.cpp), but is possibly unused?

**Comment by ReynoldsCahoon** [comment]
In the TrinityCore discussion referenced in this issue, user @FALL1N1 references getting this working in some fashion in AzerothCore.
https://github.com/TrinityCore/TrinityCore/issues/15057#issuecomment-383313037

After poking around this user's repositories, I believe the code he's referencing may be [here](https://github.com/FALL1N1/HysteriaCore/tree/master/src/server/game/VoiceChat).

**Comment by ReynoldsCahoon** [comment]
Spoke with @FALL1N1 on Discord and he gave an example of how to enable the clientside interface:

> you can test it by enabling the voice chat UI through CharacterHandler.cpp, more-specifically this line https://github.com/FALL1N1/HysteriaCore/blob/4ff15b9ed80f989b7a793637e357d928b432a75d/src/server/game/Handlers/CharacterHandler.cpp#L899
> 
> set it to '1' or 'true' within your core and the voice UI will appear, that way you can try to 'record' your voice and play it, you will see what I'm talking about :D

**Comment by yuanf225** [comment]
hey guys Is anyone making this cool feature?

**Comment by Winfidonarleyan** [comment]
No one will do it.

**Comment by Si1ker** [comment]
www.discord.com 🐱

**Comment by FALL1N1** [comment]
It's trash, better stick to Discord or TeamSpeak.

If you want it just for the cosmetics/visuals you can take a look at my repo. Voice quality is trash as the game is using old codec that cannot be changed unless you reverse the game and provide custom WoW.exe

**Comment by WarlordXecc** [comment]
If a system like this would extend from channels and be able to just be open world chat letting players speak to other players that's in a certain range of each other then I'll juicen up the bounty with another 500$.

**Comment by benjymansy123** [comment]
Would the packets for voice chat show up in a sniff?

**Comment by Nyeriah** [comment]
Voice chat was removed from the game. Iirc audio quality in 335 is awful too. There are plenty of reasons to look the other way honestly.

**Comment by FALL1N1** [comment]
@benjymansy123 
> Would the packets for voice chat show up in a sniff?

They won't unless you are sniffing a server that has coded it, otherwise the server will return empty responses (if anything at all)

**Comment by Wowkasu** [comment]
Please do not mention discord and steam, everyone is well aware of these ways. You're talking about human-to-human conversation, but that's not what's needed here, you need to hear a voice through your character from another character. It's not about quality, it's about hearing the voice from another character, just like you would hear it from Lady Silvana or Artos or Garosh. You need it to play the voice within the game, not to communicate with a friend or person. Do you understand the difference?

For example, you're walking through the game world... doing quests... and all of a sudden you hear a voice saying "help!..." "I'm dying...!" and you come because you hear another player's human voice through their character!!! it's very, very beautiful for the game!

Translated with www.DeepL.com/Translator (free version)

**Comment by FALL1N1** [comment]
> Please do not mention discord and steam, everyone is well aware of these ways. You're talking about human-to-human conversation, but that's not what's needed here, you need to hear a voice through your character from another character. It's not about quality, it's about hearing the voice from another character, just like you would hear it from Lady Silvana or Artos or Garosh. You need it to play the voice within the game, not to communicate with a friend or person. Do you understand the difference?
> 
> For example, you're walking through the game world... doing quests... and all of a sudden you hear a voice saying "help!..." "I'm dying...!" and you come because you hear another player's human voice through their character!!! it's very, very beautiful for the game!

I'm afraid that's not how Voice Chat works and worked back in the day, it's not an open world voice chat like in other games, but limited to Party, Raid and World channel(s).

**Comment by acidmanifesto** [comment]
> Please do not mention discord and steam, everyone is well aware of these ways. You're talking about human-to-human conversation, but that's not what's needed here, you need to hear a voice through your character from another character. It's not about quality, it's about hearing the voice from another character, just like you would hear it from Lady Silvana or Artos or Garosh. You need it to play the voice within the game, not to communicate with a friend or person. Do you understand the difference?
> 
> For example, you're walking through the game world... doing quests... and all of a sudden you hear a voice saying "help!..." "I'm dying...!" and you come because you hear another player's human voice through their character!!! it's very, very beautiful for the game!

You are exceeding wrong with how the voice chat work back when wotlk was retail. It did not work like that at all.

**Comment by Wowkasu** [comment]
> > Please do not mention discord and steam, everyone is well aware of these ways. You're talking about human-to-human conversation, but that's not what's needed here, you need to hear a voice through your character from another character. It's not about quality, it's about hearing the voice from another character, just like you would hear it from Lady Silvana or Artos or Garosh. You need it to play the voice within the game, not to communicate with a friend or person. Do you understand the difference?
> > For example, you're walking through the game world... doing quests... and all of a sudden you hear a voice saying "help!..." "I'm dying...!" and you come because you hear another player's human voice through their character!!! it's very, very beautiful for the game!
> 
> You are exceeding wrong with how the voice chat work back when wotlk was retail. It did not work like that at all.
I didn't write about how it was in the past, I wrote about what it was for in the long run.

**Comment by Rastrian** [comment]
https://github.com/celguar/voicechat-server

theres this implementation for ascent that can be ported to AC

**Comment by heyitsbench** [comment]
> theres this implementation for ascent that can be ported to AC

That repo's been updated for use with the TBC branch of CMaNGOS (PR pending), and includes no server code.

**Comment by FALL1N1** [comment]
> > theres this implementation for ascent that can be ported to AC
> 
> That repo's been updated for use with the TBC branch of CMaNGOS (PR pending), and includes no server code.

There are no opcode structure changes between TBC and WoTLK (at least not any major ones) therefore it should work out of the box.

https://github.com/celguar/voicechat-server is the new executable you need to "slap" into "src/server/apps" so it can build alongside with the world/auth servers (unless you want to separate it to another vps/dedi server if bandwidth is a problem).. and https://github.com/cmangos/mangos-tbc/pull/668/files is the "core" part, which handles the disabled (at the moment) opcodes alongside with their handlers.

**Comment by celguar** [comment]
Yes I tested same code with cmangos wotlk + 3.3.5 and it worked

**Comment by ReynoldsCahoon** [comment]
> I'm afraid that's not how Voice Chat works and worked back in the day, it's not an open world voice chat like in other games, but limited to Party, Raid and World channel(s).

Correct. @Foereaper and I have [created a proof of concept](https://github.com/ReynoldsCahoon/WotLK-Mumble-Positional-Voice) using the open source software Mumble that allows for proximity voice chat. This unfortunately requires an external client, but works.

If the voice packets could be manipulated to match the format expected by Mumble, you might be able to utilize something like Mumble's server software without requiring players to use third party voice client software. The benefits being obvious (nothing extra required to install, voice settings/keybinds manageable from the UI, muting and disabling of voice within the client itself).

I acknowledge this is beyond the scope of this issue, but having the server properly handle these packets in a default way gets custom functionality like this one step closer to existing.
